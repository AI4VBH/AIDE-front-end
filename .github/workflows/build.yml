name: 'Build'

on:
  pull_request:

env:
  NODE_VERSION: '16.x'

jobs:
  check-build-integrity:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
        with:
            fetch-depth: 0
      - uses: actions/setup-node@v3
        with:
          node-version: ${{ env.NODE_VERSION }}
          cache: 'npm'
      - run: npm ci
      - run: npm run build

  lint:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
        with:
            fetch-depth: 0
      - uses: actions/setup-node@v3
        with:
          node-version: ${{ env.NODE_VERSION }}
          cache: 'npm'
      - run: npm ci
      - run: sudo npx eslint './src/**/*.ts' -f json -o ./eslint.json || exit 0

      - name: Archive eslint
        uses: actions/upload-artifact@v3
        with:
          name: eslint
          path: eslint.json
