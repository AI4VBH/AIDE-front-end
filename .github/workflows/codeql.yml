# Copyright 2022 Guy’s and St Thomas’ NHS Foundation Trust
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

name: CodeQL

on:
  pull_request:
    paths:
      - 'src/**'
  push:
    branches:
      - 'develop'
      - 'main'
  workflow_dispatch:

env:
  NODE_VERSION: '16.x'

jobs:
  analyze:
    runs-on: ubuntu-latest

    permissions:
      security-events: write

    steps:
    - name: Checkout repository
      uses: actions/checkout@v3
      with:
        fetch-depth: 0
    
    - name: Setup Node.js
      uses: actions/setup-node@v3
      with:
        node-version: ${{ env.NODE_VERSION }}
        cache: 'npm'

    - name: Build Application
      run: npm ci

    - name: Initialize CodeQL
      uses: github/codeql-action/init@v2
      with:
        languages: javascript, typescript    

    - name: Perform CodeQL Analysis
      uses: github/codeql-action/analyze@v2
