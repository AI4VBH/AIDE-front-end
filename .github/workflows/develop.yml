# Copyright 2022 Guy’s and St Thomas’ NHS Foundation Trust
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

name: 'develop'

# Controls when the workflow will run
on:
  # Triggers the workflow on push events but only for the develop branch
  push:
    branches: [ develop ]

jobs:
  check_build_integrity:
    name: Check Build Integrity
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v3
      - uses: actions/setup-node@v3
        with:
          node-version: 16
      - run: npm ci
      - run: npm run build

  Deploy:
    needs: [check_build_integrity]
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v3

      - name: Configure AWS Credentials
        uses: aws-actions/configure-aws-credentials@v1
        with:
          aws-access-key-id: ${{ secrets.AWS_ACCESS_KEY }}
          aws-secret-access-key: ${{ secrets.AWS_SECRET }}
          aws-region: eu-west-2

      - name: Login to Amazon ECR Private
        id: login-ecr
        uses: aws-actions/amazon-ecr-login@v1

      - name: Set branch name
        run: echo "BRANCH_NAME=$(echo $(date +%Y.%m.%d))-develop.$(echo $(date +%-H%M))" >> $GITHUB_ENV

      - name: Build and push
        uses: docker/build-push-action@v3
        with:
          push: true
          pull: true
          tags: ${{ secrets.AWS_ECR_REGISTRY }}/aide/front-end:${{ env.BRANCH_NAME }}

      - name: Create artifacts folder and copy helm charts
        run: |
          mkdir -p "$GITHUB_WORKSPACE/artifacts/aide-frontend/helm"
          cp -a $GITHUB_WORKSPACE/helm/. $GITHUB_WORKSPACE/artifacts/aide-frontend/helm/
          echo BASE_PATH="$GITHUB_WORKSPACE/artifacts/aide-frontend" >> $GITHUB_ENV
          echo OUT_FOLDER="$GITHUB_WORKSPACE/artifacts/" >> $GITHUB_ENV

      - name: Install Octopus CLI
        uses: OctopusDeploy/install-octopus-cli-action@v1.2.1
        with:
          version: latest

      - name: Package AIDE FrontEnd
        run: |
          octo pack --id="AIDE.FrontEnd" --format="Zip" --version="${{ env.BRANCH_NAME }}" --basePath="${{ env.BASE_PATH }}" --outFolder="${{ env.OUT_FOLDER }}" --logLevel=debug

      - name: Push Package to Octopus Deploy
        uses: OctopusDeploy/push-package-action@v3.0.2
        env:
          OCTOPUS_API_KEY: ${{ secrets.OCTOPUS_API_KEY }}
          OCTOPUS_HOST: ${{ secrets.OCTOPUS_HOST }}
          OCTOPUS_SPACE: ${{ secrets.OCTOPUS_SPACE }}
        with:
          packages: |
            **/artifacts/AIDE.FrontEnd.${{ env.BRANCH_NAME }}.zip

      - name: Create Release in Octopus Deploy
        uses: OctopusDeploy/create-release-action@v2.1.0
        env:
          OCTOPUS_API_KEY: ${{ secrets.OCTOPUS_API_KEY }}
          OCTOPUS_HOST: ${{ secrets.OCTOPUS_HOST }}
          OCTOPUS_SPACE: ${{ secrets.OCTOPUS_SPACE }}
        with:
          project: 'AIDE Frontend'
          git_ref: 'main'
          channel: 'DEV'
          release_number: "${{ env.BRANCH_NAME }}"
          package_version: "${{ env.BRANCH_NAME }}"
