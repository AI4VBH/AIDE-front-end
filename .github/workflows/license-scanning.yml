name: License Finder

on:
  pull_request:
  workflow_dispatch:

env:
  NODE_VERSION: '16.x'

jobs:
  scan:
    runs-on: ubuntu-latest
    steps:
    - name: Checkout repository
      uses: actions/checkout@v3
      with:
        fetch-depth: 0

    - name: Install Ruby
      uses: ruby/setup-ruby@v1
      with:
        ruby-version: '3.0' 
        bundler-cache: true 
    - name: Install License Finder tool with Ruby
      run: gem install license_finder
    - uses: actions/setup-node@v3
      with:
        node-version: ${{ env.NODE_VERSION }}
        cache: 'npm'
    - run: npm ci
    - name: Perform License Scanning
      id: license_scanning
      run: license_finder
    - name: Check this step to see how to resolve failure with license scanning
      if: ${{ failure() && steps.license_scanning.conclusion == 'failure' }}
      run: echo "You have added a new package that needs to be manually checked to ensure it fits our license allowances. Please read the 'License Scanning' section of the 'CONTRIBUTING.MD' file to see what the next steps are."