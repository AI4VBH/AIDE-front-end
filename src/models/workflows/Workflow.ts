export interface PaginatedWorkflowsResponse {
    totalPages: number;
    totalRecords: number;
    data: WorkflowListItem[];
}

export interface WorkflowListItem {
    workflow_id: string;
    revision: number;
    name: string;
    version: string;
    description: string;
}

export interface MonaiWorkflow {
    id: string;
    workflow_id: string;
    revision: number;
    workflow: unknown;
}

export interface WorkflowError {
    type: string;
    title: string;
    status: number;
    detail: string;
    instance: string;
    traceId: string;
}
