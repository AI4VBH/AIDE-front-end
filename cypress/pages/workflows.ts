import ApiMocks from "fixtures/mockIndex";
import { AbstractPage } from "./abstractPage";
import { WorkflowData } from "data/workflows/workflows";
import { WorkflowExampleData } from "data/workflows/workflowJson";

const nextPage = ".v-data-footer__icons-after > .v-btn > .v-btn__content > .v-icon";
const errorOne =
    "Failed to validate workflow: 'router_1kjgdksgdkasgdkhasgdkhasgdhasgdjhs' is not a valid Workflow Name (source: Workflow).";
const errorTwo = "Found Task(s) without any task destinations to it: export-task-mr";
const errorThree = "Task: argo-task-mr has task destination: export-task that could not be found.";
const ErrorFour = "ERROR: Task destination export-task not found.";

export default class Workflows extends AbstractPage {
    public initPage() {
        cy.intercept("/workflows?pageNumber=1&pageSize=10", ApiMocks.WORKFLOWS).as("workflows");
        cy.visit("/workflows");
        cy.wait(["@workflows"]);
        Cypress.on("uncaught:exception", () => {
            return false;
        });
    }

    public initPageError(statuscode: number) {
        cy.intercept("/workflows?pageNumber=1&pageSize=10", { statusCode: statuscode }).as(
            "workflows",
        );
        cy.visit("/workflows");
        cy.wait(["@workflows"]);
        Cypress.on("uncaught:exception", () => {
            return false;
        });
    }

    public paginationRequestWorkflows() {
        cy.intercept("GET", `/workflows?pageNumber=2&pageSize=10`, ApiMocks.WORKFLOWS).as(
            "nextPage",
        );
        this.clickGet(nextPage);
        cy.wait(["@nextPage"]);
        Cypress.on("uncaught:exception", () => {
            return false;
        });
    }

    public assertCorrectDeleteRequest(workflow: WorkflowData) {
        cy.intercept("DELETE", `/workflows/${workflow.data[0].workflow_id}`, ApiMocks.WORKFLOWS).as(
            "Delete",
        );
        cy.intercept("GET", "/workflows?pageNumber=1&pageSize=10", ApiMocks.WORKFLOWS).as("Get");
        this.clickDataCy("workflow-delete-ok");
        Cypress.on("uncaught:exception", () => {
            return false;
        });
        cy.wait(["@Delete", "@Get"]);
    }

    public errorDeleteWorkflow(workflow: WorkflowData, statusCode: number) {
        cy.intercept("DELETE", `/workflows/${workflow.data[0].workflow_id}`, {
            statusCode: statusCode,
        }).as("Delete");
        cy.intercept("GET", "/workflows?pageNumber=1&pageSize=10", ApiMocks.WORKFLOWS).as("Get");
        this.clickDataCy("workflow-delete-ok");
        Cypress.on("uncaught:exception", () => {
            return false;
        });
        cy.wait(["@Delete"]);
    }

    public editButtonVisible() {
        [0, 1, 2, 3, 4, 5, 6, 7, 8, 9].forEach((row) => {
            cy.dataCy(`workflow-table-row-actions-${row}`).within(() => {
                cy.dataCy("workflow-edit").should("be.visible");
            });
        });
    }

    public deleteButtonVisible() {
        [0, 1, 2, 3, 4, 5, 6, 7, 8, 9].forEach((row) => {
            cy.dataCy(`workflow-table-row-actions-${row}`).within(() => {
                cy.dataCy("workflow-delete").should("be.visible");
            });
        });
    }

    public elementVisibleDataCy(tag: string) {
        cy.wait(500);
        cy.dataCy(tag).should("be.visible");
    }

    public assertTableDataCorrect(workflow: WorkflowData) {
        [0, 1, 2, 3, 4, 5].forEach((row) => {
            cy.dataCy(`workflow-table-row-name-${row}`).should(`contain`, workflow.data[row].name);
            cy.dataCy(`workflow-table-row-aetitle-${row}`).should(
                `contain`,
                workflow.data[row].ae_title,
            );
            workflow.data[row].data_origins.forEach((data_origin) => {
                cy.dataCy(`workflow-table-row-data-${row}`).should(`contain`, data_origin);
            });
            cy.dataCy(`workflow-table-row-version-${row}`).should(
                `contain`,
                workflow.data[row].version,
            );
            cy.dataCy(`workflow-table-row-description-${row}`).should(
                `contain`,
                workflow.data[row].description,
            );
        });
    }

    public elementVisibleGet(tag: string) {
        cy.get(tag).should("be.visible");
    }

    public assertTotalWorkflows(workflow: WorkflowData) {
        cy.get(".v-data-footer__pagination").should("contain", workflow.totalRecords);
    }

    public checkNextPageButtonStatus(workflow: WorkflowData) {
        if (workflow.totalRecords > 10) {
            cy.get(".v-data-footer__icons-after > .v-btn").should("be.enabled");
        } else if (workflow.totalRecords <= 10) {
            cy.get(".v-data-footer__icons-after > .v-btn").should("be.disabled");
        }
    }

    public clickDeleteButtonWorkflows(index: number) {
        cy.dataCy(`workflow-table-row-actions-${index}`).within(() => {
            cy.dataCy("workflow-delete").click();
        });
    }

    public clickEditButtonUsers(index: number) {
        cy.dataCy(`workflow-table-row-actions-${index}`).within(() => {
            cy.dataCy("workflow-edit").click();
        });
    }

    public assertEditUrl(workflow: WorkflowData) {
        cy.url().should("include", `workflow-editor/${workflow.data[0].workflow_id}`);
    }

    public assertAddUrl() {
        cy.url().should("include", `workflow-editor`);
    }

    public assertJsonError() {
        cy.get(".cm-content").clear().type("{ {");
        cy.get(".jse-message").should("be.visible");
        cy.get(".cm-content").clear().type("[]").type("{backspace}");
        cy.get(".jse-message").should("be.visible");
        cy.get(".cm-content").clear().type("{").type("{rightArrow}").type("{");
        cy.get(".jse-message").should("be.visible");
    }

    public triggerJsonValidation() {
        cy.get(".cm-content").clear().type("{ {");
    }

    public clearJson() {
        cy.get(".cm-content").clear();
    }

    public assertSaveStatus(status: string) {
        switch (status) {
            case "enabled":
                cy.dataCy("save-workflow-changes").should("be.enabled");
                break;
            case "disabled":
                cy.dataCy("save-workflow-changes").should("be.disabled");
                break;
            default:
                throw "This scenario has not been implemented. Please review";
        }
    }

    public workflowEditRequest() {
        cy.intercept(`/workflows/12345-abcde`, ApiMocks.WORKFLOW_EMPTY).as("workflows");
        cy.dataCy(`workflow-table-row-actions-0`).within(() => {
            cy.dataCy("workflow-edit").click();
        });
        cy.wait(["@workflows"]);
        Cypress.on("uncaught:exception", () => {
            return false;
        });
    }

    public workflowEditRequestError(statusCode) {
        cy.intercept(`/workflows/12345-abcde`, { statusCode: statusCode }).as("workflows");
        cy.dataCy(`workflow-table-row-actions-0`).within(() => {
            cy.dataCy("workflow-edit").click();
        });
        cy.wait(["@workflows"]);
        Cypress.on("uncaught:exception", () => {
            return false;
        });
    }

    public workflowEditRequestEmpty() {
        cy.intercept(`/workflows/12345-abcde`, ApiMocks.WORKFLOW_EMPTY).as("workflows");
        cy.dataCy(`workflow-table-row-actions-0`).within(() => {
            cy.dataCy("workflow-edit").click();
        });
        cy.wait(["@workflows"]);
        Cypress.on("uncaught:exception", () => {
            return false;
        });
    }

    public workflowDisplayedCorrectly() {
        cy.wait(500);
        cy.get(".cm-content").should("contain", "router_1");
    }

    public assertNoModal() {
        cy.get("modal").should("not.exist");
    }

    public editWorkflow() {
        cy.get(".cm-content").clear().wait(500).type("{{}}").wait(500);
    }

    public assertPut() {
        cy.intercept("PUT", `/workflows/12345-abcde`).as("put");
        this.clickDataCy("workflow-edit-confirm-ok");
        cy.wait(["@put"]);
        Cypress.on("uncaught:exception", () => {
            return false;
        });
    }

    public assertPost() {
        cy.intercept("POST", `/workflows`, { statusCode: 200 }).as("post");
        this.clickDataCy("workflow-edit-confirm-ok");
        cy.wait(["@post"]);
        Cypress.on("uncaught:exception", () => {
            return false;
        });
    }

    public returnErrorPost(statusCode: number) {
        cy.intercept("POST", `/workflows`, { statusCode: statusCode }).as("post");
        this.clickDataCy("workflow-edit-confirm-ok");
        cy.wait(["@post"]);
        Cypress.on("uncaught:exception", () => {
            return false;
        });
    }

    public returnEditError() {
        cy.intercept("PUT", `/workflows/12345-abcde`, {
            statusCode: 400,
            body: ApiMocks.WORKFLOW_ERRORS,
        }).as("error");
        this.clickDataCy("workflow-edit-confirm-ok");
        cy.wait(["@error"]);
        Cypress.on("uncaught:exception", () => {
            return false;
        });
    }

    public returnEditToastError(workflow: WorkflowExampleData, statusCode: number) {
        cy.intercept("PUT", `/workflows/12345-abcde`, {
            statusCode: statusCode,
        }).as("error");
        this.clickDataCy("workflow-edit-confirm-ok");
        cy.wait(["@error"]);
        Cypress.on("uncaught:exception", () => {
            return false;
        });
    }

    public returnAddError() {
        cy.intercept("POST", `/workflows`, { statusCode: 400, body: ApiMocks.WORKFLOW_ERRORS }).as(
            "post",
        );
        this.clickDataCy("workflow-edit-confirm-ok");
        cy.wait(["@post"]);
        Cypress.on("uncaught:exception", () => {
            return false;
        });
    }

    public errorsDisplayed() {
        cy.dataCy("error-container").should("contain", errorOne);
        cy.dataCy("error-container").should("contain", errorTwo);
        cy.dataCy("error-container").should("contain", errorThree);
        cy.dataCy("error-container").should("contain", ErrorFour);
    }

    public returnSuccess() {
        cy.intercept("PUT", `/workflows/12345-abcde`, {
            statusCode: 200,
        }).as("success");
        this.clickDataCy("workflow-edit-confirm-ok");
        cy.wait(["@success"]);
        Cypress.on("uncaught:exception", () => {
            return false;
        });
    }

    public assertWorkflowsUrl() {
        cy.url().should("eql", "http://localhost:8080/workflows");
    }

    public assertNoText() {
        cy.get(".cm-content").should("contain", "{}");
    }
}
