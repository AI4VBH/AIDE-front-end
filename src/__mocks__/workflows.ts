import { WorkflowListItem } from "@/models/workflows/Workflow";
import { rest } from "msw";

const workflows: WorkflowListItem[] = [
    {
        workflow_id: "72657913-8ba7-4bba-b04f-5de8f0ce6b29",
        name: "workflow 1",
        version: "1",
        revision: 1,
        description: "This is a workflow description",
    },
    {
        workflow_id: "92657913-8ba7-4bba-b04f-5de8f0ce6b29",
        name: "workflow 2",
        version: "2",
        revision: 1,
        description: "",
    },
    {
        workflow_id: "26657913-8ba7-4bba-b04f-5de8f0ce6b29",
        name: "workflow 3",
        version: "2.2",
        revision: 1,
        description: "This is a workflow description",
    },
];

export const workflowsHandlers = [
    rest.get(`${window.FRONTEND_API_HOST}/workflows`, (req, res, ctx) => {
        const pageNumber = req.url.searchParams.get("pageNumber") || "0";

        return res(
            ctx.json({
                totalPages: pageNumber,
                totalRecords: workflows.length,
                data: workflows,
            }),
        );
    }),
    rest.post(`${window.FRONTEND_API_HOST}/workflows`, (req, res, ctx) => {
        return res(ctx.status(201));
    }),
    rest.put(`${window.FRONTEND_API_HOST}/workflows/:workflowId`, (req, res, ctx) => {
        return res(ctx.status(200));
    }),
    rest.delete(`${window.FRONTEND_API_HOST}/workflows/:workflowId`, (req, res, ctx) => {
        return res(ctx.status(200));
    }),
];
