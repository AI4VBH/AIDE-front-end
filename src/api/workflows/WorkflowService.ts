/*
 * Copyright 2022 Guy’s and St Thomas’ NHS Foundation Trust
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { MonaiWorkflow, PaginatedWorkflowsResponse } from "@/models/workflows/Workflow";
import { createAxiosInstance, ErrorMessageMap, isResultOk } from "@/utils/axios-helpers";
import { AxiosError, AxiosResponse } from "axios";

const errorMessagesWorkflows: ErrorMessageMap = {
    get: "Something unexpected went wrong retrieving workflows",
    post: "Something unexpected went wrong creating the workflow",
    put: "Something unexpected went wrong updating the workflow",
    delete: "Something unexpected went wrong deleting the workflow",
};

const httpWorkflows = createAxiosInstance(errorMessagesWorkflows);

interface QueryParams {
    itemsPerPage: number;
    page: number;
}

export async function getAllWorkflows(query: QueryParams): Promise<PaginatedWorkflowsResponse> {
    const params = new URLSearchParams({
        pageNumber: `${query.page}`,
        pageSize: `${query.itemsPerPage}`,
    });

    const response = await httpWorkflows.get<PaginatedWorkflowsResponse>(`/workflows?${params}`);

    const defaultData = { totalPages: 0, totalRecords: 0, data: [] };

    return isResultOk(response) ? response.data : defaultData;
}

export async function getWorkflow(workflowId: string): Promise<MonaiWorkflow> {
    const response = await httpWorkflows.get<MonaiWorkflow>(`/workflows/${workflowId}`);
    return response.data;
}

export async function updateWorkflow(
    workflowId: string,
    workflow: unknown,
    orignal_workflow_name: string,
): Promise<AxiosResponse | AxiosError> {
    return httpWorkflows
        .put(`/workflows/${workflowId}`, { orignal_workflow_name, workflow })
        .catch((error) => {
            if (error) {
                return error;
            }
        });
}

export async function createWorkflow(workflow: unknown): Promise<AxiosResponse | AxiosError> {
    return httpWorkflows.post("/workflows", workflow).catch((error) => {
        if (error) {
            return error;
        }
    });
}

export async function deleteWorkflow(workflowId: string): Promise<boolean> {
    try {
        const result = await httpWorkflows.delete(`/workflows/${workflowId}`);
        return isResultOk(result);
    } catch {
        return false;
    }
}
