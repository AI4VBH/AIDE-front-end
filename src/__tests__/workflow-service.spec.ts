/*
 * Copyright 2022 Guy’s and St Thomas’ NHS Foundation Trust
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { rest } from "msw";
import { setupServer } from "msw/node";
import { createMock } from "@golevelup/ts-jest";
import { expect, test as it } from "@jest/globals";
import Vue from "vue";
import { ToastApi } from "vue-toast-notification";

import { getAllWorkflows } from "@/api/workflows/WorkflowService";
import { AxiosError } from "axios";

jest.mock("vue");

const server = setupServer();
const mockToaster = createMock<ToastApi>();
Vue.$toast = mockToaster;

describe("WorkflowService", () => {
    beforeAll(() => {
        server.listen();
    });

    afterAll(() => {
        server.close();
    });

    beforeEach(() => {
        server.resetHandlers();
        jest.resetAllMocks();
    });

    describe("getAllWorkflows", () => {
        it.each([[1, 10, "?pageNumber=1&pageSize=10"]])(
            "passes params correctly",
            async (page, itemsPerPage, query) => {
                server.use(
                    rest.get("*/workflows", (req, res, ctx) => {
                        const search = req.url.search;

                        expect(search).toBe(query);

                        return res(ctx.json({ totalPages: 0, totalRecords: 0, data: [] }));
                    }),
                );

                const result = await getAllWorkflows({ page, itemsPerPage });

                expect(result).toMatchSnapshot();
            },
        );
    });
});
