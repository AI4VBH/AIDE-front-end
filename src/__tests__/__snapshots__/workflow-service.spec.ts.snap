// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkflowService getAllWorkflows passes params correctly 1`] = `
Object {
  "data": Array [],
  "totalPages": 0,
  "totalRecords": 0,
}
`;
